---
title: "Measuring biological generality in meta-analysis: a pluralistic approach to heterogeneity quantification and stratification"
subtitle: "A step-by-step illustration"
author: "Yefeng Yang, Daniel W. A. Noble, Rebecca Spake, Alistair M. Senior, Malgorzata Lagisz, Shinichi Nakagawa"
date: "last update Nov. 2023"
output:
  rmdformats::robobook:
    code_folding: hide
    self_contained: true
    thumbnails: false
    lightbox: true
pkgdown:
  as_is: true  
bibliography: "references.bib"
csl: "nature.csl"
link-citations: yes  
---

# Contributors

Yefeng Yang, Daniel W. A. Noble, Rebecca Spake, Alistair M. Senior, Malgorzata Lagisz, Shinichi Nakagawa

# Update

Last update Nov 2023

# Preface

This step-by-step tutorial is a supplement to our paper targeted at measuring biological generality via meta-analytic heterogeneity.


## Citation

If our paper and tutorial have helped you, please cite the following paper:

> Yefeng Yang, Daniel W. A. Noble, Rebecca Spake, Alistair M. Senior, Malgorzata Lagisz, Shinichi Nakagawa. Measuring biological generality in meta-analysis: a pluralistic approach to heterogeneity quantification and stratification. EcoEvoRxiv, 2023. 

## Contact

If you have any questions, mistakes, or bug to report, please contact corresponding authors:

- Dr. Yefeng Yang

Evolution & Ecology Research Centre, EERC
School of Biological, Earth and Environmental Sciences, BEES
The University of New South Wales, Sydney, Australia

Email: yefeng.yang1@unsw.edu.au

- Professor Shinichi Nakagawa, PhD, FRSN

Evolution & Ecology Research Centre, EERC
School of Biological, Earth and Environmental Sciences, BEES
The University of New South Wales, Sydney, Australia

Email: s.nakagawa@unsw.edu.au

```{r global options, include=FALSE}
library(knitr)
library(rmdformats)
## Global options
# options(max.print = "75")
knitr::opts_chunk$set(
  echo = FALSE, cache = FALSE, prompt = FALSE,
  tidy = FALSE, comment = NA,
  message = FALSE, warning = FALSE
)
opts_knit$set(width = 75)

```

# Set-up 

Our illustrations use R statistical software and existing R packages, which you will first need to download and install.

If you do not have it on your machine, first install `R` ([download](https://cran.r-project.org/)). We recommend also downloading `RStudio`, a popular integrated development environment for coding with `R`, created by a company named posit ([download](https://posit.co/products/open-source/rstudio/)). 

After installing `R`, you must install several packages that contain necessary functions for performing the analyses in this tutorial. If the packages are archived in CRAN, use `install.packages()` to install them. For example, to install the `metafor` package (a common meta-analysis package), you can execute `install.packages("metafor")` in the console (bottom left pane of `R Studio`). To install packages that are not on CRAN and archived in Github repositories, execute `devtools::install_github()`. For example, to install `orchaRd` (a meta-analysis visualization package) from Github repository, execute `devtools::install_github("daniel1noble/orchaRd", force = TRUE)`.

The package list is as follows:

```{r load packages, cache = FALSE, echo=T}
# Install and load necessary library
pacman::p_load(tidyverse, 
               here,
               DT,
               janitor,
               ggpubr,
               readxl, 
               metafor,
               ggplot2,
               ggsignif,
               visdat,
               cowplot,
               patchwork,
               pander,
               rotl,
               ape,
               ggstance,
               ggtree,
               RColorBrewer,
               wesanderson
               )

```

# Custom function

We provide helper functions necessary for our illustrations. To use them, you need to `source` them. You also can paste the source code into the console, and hit "Enter" to let `R` "learn" these custom functions.

```{r function, echo = T}
source(here("function","custom_function.R"))
```


# Load data and pre-process

Hodkinson and colleagues  conducted a network meta-analysis to assess the efficacy of different self-management interventions (multidisciplinary case management, regularly supported self-management, and minimally supported self-management) in enhancing the quality of life among asthma patients.

```{r data, warning=FALSE, echo = T}
#load data
dat_Risely <- read.csv(here("data","Risely_2018_data.csv"))

#subset observations on infection status
status <- subset(dat_Risely, infection.measure=="Infection status") ##subset observations on infection status

#t1 <- status %>% DT::datatable()
```

We also construct the phylogenetic tree to allow for decomposing heterogeneity at both species and phylogeny levels.

```{r tree, warning=FALSE, echo = T}
#check species we have
#length(unique(status$species_latin)) #25 unique species names, if no misspelling

#find Open Tree Taxonomy (OTT) IDs for each species
taxa <- tnrs_match_names(names = unique(status$species_latin))

#rough check
#length(taxa$unique_name) # 25 unique species names, which is aligned with the species names in the dataset
#tabyl(taxa,approximate_match) # 0 approximate match

#check whether all otts occur in the synthetic tree
ott_in_tree <- ott_id(taxa)[is_in_tree(ott_id(taxa))]  # all good
#length(ott_in_tree) # 25

#make phylo tree
tree <- tol_induced_subtree(ott_ids = ott_id(taxa))

tree$tip.label <- strip_ott_ids(tree$tip.label, remove_underscores = TRUE)

# test whether a tree is binary
#is.binary(tree)  

#decapitalise species names to match with the search string names in taxa
status <- status %>% mutate(search_string = tolower(species_latin))  

#align data
status <- left_join(status, dplyr::select(taxa, search_string, unique_name, ott_id), by = "search_string")  

#create the variables of spp and phylo
status <- status %>% mutate(spp = search_string, phylo = unique_name)


#estimate branch lengths using APE
tree2 <- compute.brlen(tree, method = "Grafen", power = 1)
##create correlation matrix
phylo_corr <- vcv(tree2, corr=TRUE)
plot(tree2, cex = .8, label.offset = .1, no.margin = TRUE)
```


# Fit data

We fit a phylogenetic multilevel model as shown in Equation 19 in the main text to the data. Doing so, we can stratify the heterogeneity at different strata: within-study level (`obs_id`), between-study level (`study_id`), species level (`spp`), and phylogeny level (`phylo`).

$$
ES_{[i]} = \mu + u_{species[k]} + u_{phylogeny[k]} + \mu_{between[j]} + \mu_{within[j]} + e_{[i]}, (19)
$$
```{r model fit, echo = T, warning=FALSE}
#unify the name of key variables
dat <- status
dat$obs_id <- 1:nrow(dat)
dat$study_id <- dat$id
# run the model
mod_status <- rma.mv(g, var.g, random = list(~ 1 | obs_id, ~1 | study_id, ~ 1 | spp, ~ 1 | phylo), R = list(phylo = phylo_corr), data = dat)

summary(mod_status)
```

# Heterogeneity quantification and stratification

We quantify and stratify the heterogeneity using the formulas provided in the main text. All functions used are integrated into `orchard` package @nakagawa2021orchard.

__Typical sampling variance (which captures the statistical noise of the data, but is rarely reported in the current meta-analytic practice): __

```{r Vbar, echo = T, warning=FALSE}
sigma2_v(mod_status)
```

__Unstandardised heterogeneity metrics $\sigma^2$:__ 

```{r sigma2, warning=FALSE}
t2 <- data.frame(stratum = c("total", mod_status$s.names),
           sigma2 = c(sum(mod_status$sigma2), mod_status$sigma2)) %>% dfround(3)

t2$stratum <- c("total", "within-study", "between-study", "species", "phylogeny")
rownames(t2) <- NULL

t2 %>% DT::datatable()
```

__Variance-standardised heterogeneity metrics $I^2$:__

```{r I2, echo = T, warning=FALSE}
t3 <- data.frame(stratum = c("total", mod_status$s.names),
           I2 = i2_ml(mod_status)) %>% dfround(1)

t3$stratum <- c("total", "within-study", "between-study", "species", "phylogeny")
rownames(t3) <- NULL

t3 %>% DT::datatable()
```

__Mean-standardised heterogeneity metrics $CV^2$:__

```{r CV, echo = T, warning=FALSE}
t4 <- data.frame(stratum = c("total", mod_status$s.names),
           CV = cv_ml(mod_status)) %>% dfround(1)

t4$stratum <- c("total", "within-study", "between-study", "species", "phylogeny")
rownames(t4) <- NULL

t4 %>% DT::datatable()
```

__Mean-standardised heterogeneity metrics $CV^2$:__

```{r M, echo = T, warning=FALSE}
t5 <- data.frame(stratum = c("total", mod_status$s.names),
           M1 = m1_ml(mod_status)) %>% dfround(1)

t5$stratum <- c("total", "within-study", "between-study", "species", "phylogeny")
rownames(t5) <- NULL

t5 %>% DT::datatable()
```

__Let's visualize all heterogeneity measures:__

```{r visualization, echo = T, warning=FALSE}
# Visualize heterogeneity
## make dataframe
h_status <- h.calc(mod_status)
h_status$levels <- rownames(h_status)  
h_status$levels <- dplyr::recode(h_status$levels, "Total" = "Total",  "obs_id" = "Within",  "study_id" = "Between", "spp" = "Spp", "phylo" = "Phylo")
h_status$levels <- as.factor(h_status$levels)
h_status$levels <- factor(h_status$levels, levels = c("Total", "Phylo", "Spp", "Between", "Within"))

## plot
## sigma
p.sigma <- ggplot(h_status, aes(levels, sigma2s)) +
  geom_col(alpha = 1, color = wes_palette('GrandBudapest1', 4, type = 'discrete')[1], fill = wes_palette('GrandBudapest1', 4, type = 'discrete')[1]) +
  labs(y = expression("Variance"), x = "Strata" , title = "Unstandardised heterogeneity metrics") + 
  theme_bw() +
  theme(legend.background = element_blank(),
        axis.text = element_text(size = 12, color = "black"),
        axis.title = element_text(size = 12, color = "black")
        ) 

# I2
p.I2 <- ggplot(h_status, aes(levels, I2s_Shinichi/100)) +
  geom_col(alpha = 1, color = wes_palette('GrandBudapest1', 4, type = 'discrete')[2], fill = wes_palette('GrandBudapest1', 4, type = 'discrete')[2]) +
  scale_y_continuous(labels = scales::percent_format()
                     ) +
  labs(y = expression(paste(italic(I)[]^2)), x = "Strata" , title = "Source of heterogeneity") + 
  theme_bw() +
  theme(legend.background = element_blank(),
        axis.text = element_text(size = 12, color = "black"),
        axis.title = element_text(size = 12, color = "black")
        ) 

# CV
p.CV <- ggplot(h_status, aes(levels, CVs)) +
  geom_col(alpha = 1, color = wes_palette('GrandBudapest1', 4, type = 'discrete')[3], fill = wes_palette('GrandBudapest1', 4, type = 'discrete')[3]) +
  labs(y = expression(paste(italic(CV)[])), x = "Strata" , title = "Magnitude of heterogeneity") + 
  scale_y_continuous(labels = scales::number_format(accuracy = 0.01)) + # https://stackoverflow.com/questions/38722202/how-do-i-change-the-number-of-decimal-places-on-axis-labels-in-ggplot2
  theme_bw() +
  theme(legend.background = element_blank(),
        axis.text = element_text(size = 12, color = "black"),
        axis.title = element_text(size = 12, color = "black")
        )   
  

# M
p.M <- ggplot(h_status, aes(x = levels, y = Ms)) +
  geom_col(alpha = 1, color = wes_palette('GrandBudapest1', 4, type = 'discrete')[4], fill = wes_palette('GrandBudapest1', 4, type = 'discrete')[4]) +
  labs(y = expression(paste(italic(M)[])), x = "Strata" , title = "Magnitude of heterogeneity") + 
  scale_y_continuous(labels = scales::number_format(accuracy = 0.01)) + # https://stackoverflow.com/questions/38722202/how-do-i-change-the-number-of-decimal-places-on-axis-labels-in-ggplot2
  theme_bw() +
  theme(legend.background = element_blank(),
        axis.text = element_text(size = 12, color = "black"),
        axis.title = element_text(size = 12, color = "black")
        )  




#png(filename = "all.png", width = 9, height = 8, units = "in", type = "windows", res = 400)
p.sigma + p.I2 + p.CV + p.M + plot_layout(ncol =2, nrow = 2) + plot_annotation(tag_levels = "A") & theme(plot.tag = element_text(face = "bold"))
#dev.off()
```


# Predictive distributions: visualising heterogeneity

Next, we illustrate how prediction intervals (PIs) combined with ‘prediction distributions’ (PDs) can be used to understand effect size heterogeneity and generality: 

```{r PD, echo = T, warning=FALSE}
############################################		
######## Simulate some meta-analytic data
############################################

set.seed(98)

# Parameters	
mu = 0.5  									   # Mean effect size. For example SMD. This is the mean contrast between two groups
tau = 0.1  									   # Between-study variance
obs = 0.6                                      # Within study varaince 
n = 50                                       # Number of studies
k = 100                                      # Number of effect sizes
vi = rgamma(k, shape = 0.5, scale = 0.1)	   # Effect size sampling variance. Sample from a gamma distribution

# Simulated effect size data
yi = mu + rep(rnorm(n, 0, sd = sqrt(tau)), k/n) + rnorm(k,0, obs) + rnorm(k, 0, sd = sqrt(vi))

# Create data
dat <- data.frame(yi = yi, vi = vi, study = rep(1:n, length = k), within = 1:k)

# Model
mod_sim <- rma.mv(yi, vi, random = list(~ 1 | study, ~1|within), data = dat, method = "REML")


# Get the necessary data
pred_dens_data <- pred_dist_data(mod_sim)[-2,]
pred_dens_data$group[1] <- "Between-study" 
pred_dens_data$group[2] <- "Total"
	
# Calculate what proportion of effect sizes are beyond a biological threshold (in this case SMD = 0.2) for each of the two levels.
pred_dens_data$threshold <- propT_beyond(pred_dens_data, df = mod_sim$k.all - 1, threshold = mod_sim$ci.lb, tail = "above")

mod_resultsdata <- mod_results(mod_sim, mod = "1", group = "study")$mod_table
# Orchardplot
sim_orchard <- orchard_plot(mod_sim, mod = "1", xlab = "Effect Size", group = "study", angle = 45, k.pos = "left") + 
	        ylim(-1.5, 2.5) +  
					geom_abline(intercept = mod_resultsdata$lowerPR, slope = 0, linetype = 2,  colour = "red") + 
					geom_abline(intercept = mod_resultsdata$upperPR, slope = 0, linetype = 2,  colour = "red") + 
     				theme(plot.margin = unit(c(0,0,0,0), 'lines'),
					  axis.title = element_text(size = 15, face = "bold"),
					  axis.text = element_text(size = 12),
					  axis.text.y = element_blank(),
					  axis.ticks.y = element_blank())


# Now create the density plot
	# t distribution
density_between <- data.frame(x = seq(-1.5,2.5,length.out = 1000)) %>% 
	  mutate(density = c(pred_Tdistri(x, df=mod_sim$k.all-1, m=pred_dens_data$mean[1], sd=pred_dens_data$sd[1]))) %>%
	  mutate(group = rep("Between-study", length(x)))
	density_total <- data.frame(x = seq(-1.5,2.5,length.out = 1000)) %>% 
	  mutate(density = c(pred_Tdistri(x, df=mod_sim$k.all-1, m=pred_dens_data$mean[2], sd=pred_dens_data$sd[2]))) %>%
	  mutate(group = rep("Total", length(x)))
	#density_dat <- rbind(density_between, density_total)
	#density_dat$group <- as.factor(density_dat$group)
	#density_dat$group <- factor(density_dat$group, levels = c("Total", "Between-study"))
	
	
dens_plot <- ggplot() +
	  geom_area(data = density_between, aes(x = x, y = density, fill = "Between-study"),alpha = 0.3) +
	  geom_area(data = density_total, aes(x = x, y = density, fill = "total"),alpha = 0.3) +
	  scale_fill_manual(values = c("#D95F02", "#1B9E77"), labels = c(bquote("Total:"~paste(italic(sigma)[total]^2)~"= 0.37,"~paste(italic(I)[total]^2)~"= 100%,"~paste(italic(M)[total])~"= 0.46"), bquote("Between-study:"~paste(italic(sigma)[between]^2)~"= 0.08,"~paste(italic(I)[between]^2)~"= 20%,"~paste(italic(M)[between])~"= 0.21"))) + 
	  stat_function(fun = pred_Tdistri_shaded, args = list(df = mod_sim$k.all - 1, m = pred_dens_data$mean[1], sd = pred_dens_data$sd[1], threshold = mod_sim$ci.lb),
	                geom = "area", fill = "#D95F02", alpha = 0.5) +
	  stat_function(fun = pred_Tdistri_shaded, args = list(df = mod_sim$k.all - 1, m = pred_dens_data$mean[2], sd = pred_dens_data$sd[2], threshold = mod_sim$ci.lb),
	                geom = "area", fill = "#1B9E77",  alpha = 0.5) +
	  theme_bw() + xlim(-1.5, 2.5) + labs(fill = "Stratum") +
	  geom_vline(xintercept = 0, linetype = 2,  colour = "black") +
	  geom_vline(xintercept = mod_resultsdata$lowerPR, linetype = 2,  colour = "red") + 
	  geom_vline(xintercept = mod_resultsdata$upperPR, linetype = 2,  colour = "red") + 
	  annotate("text", x = pred_dens_data$mean, y = c(0.8, 0.5), label = paste0(pred_dens_data$threshold, "%"), size = 5) + 
	  theme(legend.position= c(0, 1), 
	        legend.justification = c(0, 1), 
	        legend.background = element_blank(),
	        legend.text = element_text(size = 10),
	        legend.title = element_text(size = 15, face = "bold"),
	        axis.text = element_blank(),
	        axis.title = element_blank(),
	        axis.ticks = element_blank(),
	        plot.margin = unit(c(0,0,0,0), 'lines')) 
	
## normal distribution	
dens_plot2 <- ggplot(data = data.frame(x=c(-1.5,2.5)), aes(x = x, color = group, fill = group)) +
	        pred_distri(pred_dens_data) + 
	        scale_fill_manual(values = c("#D95F02", "#1B9E77"), labels = c(bquote("Total:"~paste(italic(sigma)[total]^2)~"= 0.37,"~paste(italic(I)[total]^2)~"= 100%,"~paste(italic(M)[total])~"= 0.46"), bquote("Between-study:"~paste(italic(sigma)[between]^2)~"= 0.08,"~paste(italic(I)[between]^2)~"= 20%,"~paste(italic(M)[between])~"= 0.21"))) + 
	        scale_color_manual(values = c("#D95F02", "#1B9E77"), labels = c(bquote("Total:"~paste(italic(sigma)[total]^2)~"= 0.37,"~paste(italic(I)[total]^2)~"= 100%,"~paste(italic(M)[total])~"= 0.46"), bquote("Between-study:"~paste(italic(sigma)[between]^2)~"= 0.08,"~paste(italic(I)[between]^2)~"= 20%,"~paste(italic(M)[between])~"= 0.21"))) +
	        stat_function(fun = pred_distri_shaded, args = list(m = pred_dens_data$mean[1], sd = pred_dens_data$sd[1], threshold = mod_sim$ci.lb),
	                geom = "area", fill = "#1B9E77", color = "#1B9E77", alpha = 0.6) +
	        stat_function(fun = pred_distri_shaded, args = list(m = pred_dens_data$mean[2], sd = pred_dens_data$sd[2], threshold = mod_sim$ci.lb),
	                geom = "area", fill = "#D95F02", color = "#D95F02", alpha = 0.6) +
	         #theme(legend.text.align = 0) +
					 theme_bw() + xlim(-1.5, 2.5) + labs(fill = "Stratum", color = "Stratum") +
					 geom_vline(xintercept = 0, linetype = 2,  colour = "black") +
					 geom_vline(xintercept = mod_resultsdata$lowerPR, linetype = 2,  colour = "red") + 
					  geom_vline(xintercept = mod_resultsdata$upperPR, linetype = 2,  colour = "red") + 
					 annotate("text", x = pred_dens_data$mean, y = c(0.8, 0.5), label = paste0(pred_dens_data$threshold, "%"), size = 5) + 
					 theme(legend.position= c(0, 1), 
			  				legend.justification = c(0, 1), 
			  				legend.background = element_blank(),
							legend.text = element_text(size = 10),
							legend.title = element_text(size = 15, face = "bold"),
			  				axis.text = element_blank(),
			  				axis.title = element_blank(),
			  				axis.ticks = element_blank(),
			  				plot.margin = unit(c(0,0,0,0), 'lines')) 

# Lets try plotting the orchard plot and the density plot together
#ggsave("sim_orchard.png", width = 8, height = 8, dpi = 300)
ggarrange(dens_plot, sim_orchard, heights = c(2, 1),
          ncol = 1, nrow = 2, align = "v", common.legend = FALSE) 

```

Effect size data are simulated assuming two effect sizes were collected from a total of `n` = 50 studies, (`k` = 100), with a $\sigma_{between}^2$ = 0.1, $\sigma_{within}^2$ = 0.6 and an overall meta-analytic mean, u, of 0.5. Red dashed lines are the upper and lower 95% PI, black dashed line the ‘null’ effect. The orchard plot 30 displays the overall meta-analytic mean, 95% confidence interval (CI) and 95% PI. The PDs were constructed using *t*-distribution with `k-1` degrees of freedom, u as location parameter, and total or between-study variance along with sampling variance of around u as scale parameter (see Equation 11). The percentage of effect sizes beyond a given threshold (i.e., the lower 95% CI) are provided.

# License  

This documented is licensed under the following license: [CC Attribution-Noncommercial-Share Alike 4.0 International](https://creativecommons.org/licenses/by-nc-sa/4.0/deed.en).  

# Software and package versions  

```{r versions}
sessionInfo() %>% pander()
```

# References  
